File: src/logging.js
const { C_RESET, C_GREEN, C_RED, C_YELLOW, C_CYAN, C_DIM, C_BOLD } = require("./config");
const { mutateDb, randomId, now } = require("./store");
const { maskKey, getApiKeys } = require("./user-utils");
const { normalizeUsage } = require("./utils");

function trimStoredRecords(db, key, maxRecords = 50000) {
  if (db[key].length > maxRecords) {
    db[key] = db[key].slice(db[key].length - maxRecords);
  }
}

function logRequestToTerminal({ method, endpoint, status, ok, stream, durationMs, userName, model, providerName, errorCode, errorMessage, promptTokens, completionTokens, finishReason }) {
  const ts = new Date().toLocaleString("zh-CN", { hour12: false, timeZone: "Asia/Shanghai" });
  const statusColor = ok ? C_GREEN : C_RED;
  const statusLabel = ok ? "OK" : "FAIL";
  const methodStr = `${C_CYAN}${method || "?"}${C_RESET}`;
  const endpointStr = `${C_BOLD}${endpoint || "?"}${C_RESET}`;
  const statusStr = `${statusColor}${C_BOLD}${status || "?"} ${statusLabel}${C_RESET}`;
  const durationStr = `${C_DIM}${durationMs || 0}ms${C_RESET}`;
  const userStr = userName ? `${C_DIM}${userName}${C_RESET}` : "";
  const modelStr = model ? `${C_YELLOW}${model}${C_RESET}` : "";
  const providerStr = providerName ? `${C_DIM}via ${providerName}${C_RESET}` : "";
  const streamStr = stream ? `${C_DIM}[stream]${C_RESET}` : "";
  const tokenStr = (promptTokens || completionTokens) ? `${C_DIM}tokens=${promptTokens}+${completionTokens}${C_RESET}` : "";
  const finishStr = finishReason ? `${C_YELLOW}finish=${finishReason}${C_RESET}` : "";

  let reasonStr = "";
  if (!ok) {
    const reason = errorMessage || errorCode || `HTTP ${status}`;
    reasonStr = ` ${C_RED}reason="${reason}"${C_RESET}`;
  }

  const parts = [
    `${C_DIM}[${ts}]${C_RESET}`,
    methodStr,
    endpointStr,
    statusStr,
    durationStr,
    userStr,
    modelStr,
    providerStr,
    streamStr,
    tokenStr,
    finishStr,
    reasonStr
  ].filter(Boolean);

  console.log(parts.join(" "));
}

function recordRequestLog({
  userId,
  userName,
  username,
  apiKeyId,
  apiKeyName,
  apiKeyPreview,
  providerId,
  providerName,
  model,
  upstreamModel,
  endpoint,
  method,
  status,
  ok,
  stream,
  durationMs,
  usage,
  errorCode,
  errorMessage,
  finishReason
}) {
  const normalized = normalizeUsage(usage) || {
    promptTokens: 0,
    completionTokens: 0,
    totalTokens: 0,
    cachedTokens: 0,
    cacheCreationTokens: 0,
    cacheMissTokens: 0,
    reasoningTokens: 0
  };

  logRequestToTerminal({
    method, endpoint, status, ok, stream, durationMs,
    userName: userName || username,
    model, providerName, errorCode, errorMessage,
    promptTokens: normalized.promptTokens,
    completionTokens: normalized.completionTokens,
    finishReason
  });

  mutateDb((db) => {
    if (!Array.isArray(db.requestLogs)) db.requestLogs = [];
    const user = db.users.find((item) => item.id === userId);
    const apiKeyRecord = user && apiKeyId
      ? getApiKeys(user).find((item) => item.id === apiKeyId)
      : null;
    const timestamp = now();
    if (apiKeyRecord) {
      apiKeyRecord.lastUsedAt = timestamp;
      apiKeyRecord.updatedAt = timestamp;
    }
    db.requestLogs.push({
      id: randomId("req"),
      userId,
      userName: String(userName || user?.name || "").trim(),
      username: String(username || user?.username || "").trim(),
      apiKeyId: String(apiKeyId || "").trim(),
      apiKeyName: String(apiKeyName || apiKeyRecord?.name || "").trim(),
      apiKeyPreview: String(apiKeyPreview || maskKey(apiKeyRecord?.key || "")).trim(),
      providerId,
      providerName: String(providerName || "").trim(),
      model: String(model || "").trim() || "unknown",
      upstreamModel: String(upstreamModel || "").trim(),
      endpoint: String(endpoint || "").trim(),
      method: String(method || "").trim().toUpperCase(),
      status: Number(status) || 0,
      ok: Boolean(ok),
      stream: Boolean(stream),
      durationMs: Number(durationMs) || 0,
      promptTokens: normalized.promptTokens,
      completionTokens: normalized.completionTokens,
      totalTokens: normalized.totalTokens,
      cachedTokens: normalized.cachedTokens,
      cacheCreationTokens: normalized.cacheCreationTokens,
      cacheMissTokens: normalized.cacheMissTokens,
      reasoningTokens: normalized.reasoningTokens,
      errorCode: String(errorCode || "").trim(),
      errorMessage: String(errorMessage || "").trim().slice(0, 500),
      timestamp
    });
    trimStoredRecords(db, "requestLogs");
  });
}

module.exports = {
  logRequestToTerminal,
  recordRequestLog,
  trimStoredRecords
};

File: backend/logging/logging.go
package logging

import (
	"fmt"
	"net/http"
	"time"

	"sapi/auth"
	"sapi/device"
	"sapi/ippure"
	"sapi/models"
	"sapi/store"
	"sapi/utils"
)

var (
	Reset  = "\033[0m"
	Green  = "\033[32m"
	Red    = "\033[31m"
	Yellow = "\033[33m"
	Cyan   = "\033[36m"
	Dim    = "\033[2m"
	Bold   = "\033[1m"
)

func logRequestToTerminal(method, endpoint string, status int, ok, stream bool, durationMs int, userName, model, providerName, errorCode, errorMessage string, promptTokens, completionTokens int, finishReason string) {
	loc, _ := time.LoadLocation("Asia/Shanghai")
	ts := time.Now().In(loc).Format("2006-01-02 15:04:05")

	statusColor := Green
	statusLabel := "OK"
	if !ok {
		statusColor = Red
		statusLabel = "FAIL"
	}

	parts := []string{
		Dim + "[" + ts + "]" + Reset,
		Cyan + method + Reset,
		Bold + endpoint + Reset,
		fmt.Sprintf("%s%s%d %s%s", statusColor, Bold, status, statusLabel, Reset),
		fmt.Sprintf("%s%dms%s", Dim, durationMs, Reset),
	}
	if userName != "" {
		parts = append(parts, Dim+userName+Reset)
	}
	if model != "" {
		parts = append(parts, Yellow+model+Reset)
	}
	if providerName != "" {
		parts = append(parts, Dim+"via "+providerName+Reset)
	}
	if stream {
		parts = append(parts, Dim+"[stream]"+Reset)
	}
	if promptTokens > 0 || completionTokens > 0 {
		parts = append(parts, fmt.Sprintf("%stokens=%d+%d%s", Dim, promptTokens, completionTokens, Reset))
	}
	if finishReason != "" {
		parts = append(parts, Yellow+"finish="+finishReason+Reset)
	}
	if !ok {
		reason := errorMessage
		if reason == "" {
			reason = errorCode
		}
		if reason == "" {
			reason = fmt.Sprintf("HTTP %d", status)
		}
		parts = append(parts, Red+"reason=\""+reason+"\""+Reset)
	}

	out := parts[0]
	for _, p := range parts[1:] {
		out += " " + p
	}
	fmt.Println(out)
}

func RecordRequestLog(params RequestLogParams) {
	normalized := normalizeUsageSimple(params.Usage)

	logRequestToTerminal(
		params.Method, params.Endpoint, params.Status, params.OK, params.Stream, params.DurationMs,
		params.UserName, params.Model, params.ProviderName, params.ErrorCode, params.ErrorMessage,
		normalized.PromptTokens, normalized.CompletionTokens, params.FinishReason,
	)

	store.AppendRequestLog(models.RequestLog{
		ID:                  auth.RandomID("req"),
		UserID:              params.UserID,
		UserName:            params.UserName,
		Username:            params.Username,
		APIKeyID:            params.APIKeyID,
		APIKeyName:          params.APIKeyName,
		APIKeyPreview:       params.APIKeyPreview,
		ProviderID:          params.ProviderID,
		ProviderName:        params.ProviderName,
		Model:               params.Model,
		UpstreamModel:       params.UpstreamModel,
		Endpoint:            params.Endpoint,
		Method:              params.Method,
		Status:              params.Status,
		OK:                  params.OK,
		Stream:              params.Stream,
		DurationMs:          params.DurationMs,
		PromptTokens:        normalized.PromptTokens,
		CompletionTokens:    normalized.CompletionTokens,
		TotalTokens:         normalized.TotalTokens,
		CachedTokens:        normalized.CachedTokens,
		CacheCreationTokens: normalized.CacheCreationTokens,
		CacheMissTokens:     normalized.CacheMissTokens,
		ReasoningTokens:     normalized.ReasoningTokens,
		ErrorCode:           params.ErrorCode,
		ErrorMessage:        params.ErrorMessage,
		ClientIPInfo:        ippure.LookupRequest(params.Request),
		ClientDevice:        device.FromRequest(params.Request),
		RequestContent:      params.RequestContent,
		Timestamp:           store.Now(),
	})
}

type RequestLogParams struct {
	UserID         string
	UserName       string
	Username       string
	APIKeyID       string
	APIKeyName     string
	APIKeyPreview  string
	ProviderID     string
	ProviderName   string
	Model          string
	UpstreamModel  string
	Endpoint       string
	Method         string
	Status         int
	OK             bool
	Stream         bool
	DurationMs     int
	Usage          interface{}
	ErrorCode      string
	ErrorMessage   string
	FinishReason   string
	RequestContent map[string]interface{}
	Request        *http.Request
}

type simpleUsage struct {
	PromptTokens        int
	CompletionTokens    int
	TotalTokens         int
	CachedTokens        int
	CacheCreationTokens int
	CacheMissTokens     int
	ReasoningTokens     int
}

func normalizeUsageSimple(usage interface{}) simpleUsage {
	normalized := utils.NormalizeUsage(usage)
	if normalized == nil {
		return simpleUsage{}
	}
	return simpleUsage{
		PromptTokens:        normalized.PromptTokens,
		CompletionTokens:    normalized.CompletionTokens,
		TotalTokens:         normalized.TotalTokens,
		CachedTokens:        normalized.CachedTokens,
		CacheCreationTokens: normalized.CacheCreationTokens,
		CacheMissTokens:     normalized.CacheMissTokens,
		ReasoningTokens:     normalized.ReasoningTokens,
	}
}
